src/nbl/video/CVulkanLogicalDevice.h

```diff
@@ -542,74 +530,105 @@ class CVulkanLogicalDevice final : public ILogicalDevice
+            {
+                auto& writeAS = vk_writeDescriptorSetAS[i];
+                writeAS = { VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET_ACCELERATION_STRUCTURE_KHR, nullptr };
+                writeAS.accelerationStructureCount = pDescriptorWrites[i].count;
+
+                vk_writeDescriptorSets[i].pNext = &writeAS;
+
+                writeAS.pAccelerationStructures = reinterpret_cast<VkAccelerationStructureKHR*>(vk_accelerationStructures.size());
                 for (uint32_t j = 0u; j < pDescriptorWrites[i].count; ++j)
+                    vk_accelerationStructures.push_back({});
+            } break;
+
+            default:
+                assert(!"Invalid code path.");
+            }
+        }
+
+        for (uint32_t i = 0u; i < descriptorWriteCount; ++i)
+        {
+            switch (pDescriptorWrites[i].info->desc->getTypeCategory())
+            {
+            case asset::IDescriptor::E_CATEGORY::EC_BUFFER:
+            {
+                vk_writeDescriptorSets[i].pBufferInfo = reinterpret_cast<size_t>(vk_writeDescriptorSets[i].pBufferInfo) + vk_bufferInfos.data();
+
+                const auto* infoSrc = pDescriptorWrites[i].info;
+                auto* infoDst = const_cast<VkDescriptorBufferInfo*>(vk_writeDescriptorSets[i].pBufferInfo);
+                for (uint32_t j = 0; j < pDescriptorWrites[i].count; ++j, ++infoSrc, ++infoDst)
                 {
-                    if (pDescriptorWrites[i].info[j].info.buffer.size)
+                    if (infoSrc->info.buffer.size)
                     {
-                        vk_bufferViews[bufferViewOffset + j] = static_cast<const CVulkanBufferView*>(pDescriptorWrites[i].info[j].desc.get())->getInternalObject();
+                        infoDst->buffer = static_cast<const CVulkanBuffer*>(infoSrc->desc.get())->getInternalObject();
+                        infoDst->offset = infoSrc->info.buffer.offset;
+                        infoDst->range = infoSrc->info.buffer.size;
                     }
-                    else
+                }
+            } break;
+
+            case asset::IDescriptor::E_CATEGORY::EC_IMAGE:
+            {
+                vk_writeDescriptorSets[i].pImageInfo = reinterpret_cast<size_t>(vk_writeDescriptorSets[i].pImageInfo) + vk_imageInfos.data();
+
+                const auto* infoSrc = pDescriptorWrites[i].info;
+                auto* infoDst = const_cast<VkDescriptorImageInfo*>(vk_writeDescriptorSets[i].pImageInfo);
+                for (uint32_t j = 0; j < pDescriptorWrites[i].count; ++j, ++infoSrc, ++infoDst)
+                {
+                    if (infoSrc->info.image.imageLayout != asset::IImage::E_LAYOUT::EL_UNDEFINED)
                     {
-                        vk_bufferViews[bufferViewOffset + j] = dummyBufferView;
+                        VkSampler vk_sampler = VK_NULL_HANDLE;
+                        if (infoSrc->info.image.sampler && (infoSrc->info.image.sampler->getAPIType() == EAT_VULKAN))
```

Comment:
hoist the check for device compatibility into `ILogicalDevice`

Reply:
Just to do these checks we would have to do all this iterate over the `pDescriptorWrites` in `ILogicalDevice::updateDescriptorSets` again, worth it?

Reply:
What you're doing is not valid anyway
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkUpdateDescriptorSets.html

you can't just set the `vk_sampler` into `VK_NULL_HANDLE` and use that...

for that you'd need the `nullDescriptor` feature
https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#features-nullDescriptor
and this is a robustness2 feature, and robustness causes overheads.

Reply:
I'd do the validation in `validateWrites`, but probably only in `NBL_DEBUG` configs

here you'd assume (or assert) that the sampler is valid and compatible with the device
and also assert that there are no immutable samplers if the `info.image.sampler` is not null

Reply:
Resolved.